File: app/src/main/kotlin/com/youcoded/app/runtime/MenuAnswerLock.kt
package com.youcoded.app.runtime

/**
 * One device at a time may answer a session's Claude Code menu by verified
 * navigation (arrows, then Enter). Port of desktop's src/main/menu-answer-lock.ts
 * (review F4, 2026-09-24): two devices clicking together could otherwise mix one
 * device's arrows with the other's Enter. A lease, so a device that dies
 * mid-answer cannot lock the session forever.
 */
class MenuAnswerLock(private val now: () -> Long = System::currentTimeMillis) {
    private data class Held(val holder: String, val until: Long)
    private val held = HashMap<String, Held>()

    @Synchronized
    fun acquire(sessionId: String, holder: String): Boolean {
        val cur = held[sessionId]
        if (cur != null && cur.holder != holder && cur.until > now()) return false
        held[sessionId] = Held(holder, now() + LEASE_MS)
        return true
    }

    @Synchronized
    fun release(sessionId: String, holder: String) {
        if (held[sessionId]?.holder == holder) held.remove(sessionId)
    }

    /** The bridge entry point: "acquire" → granted?, anything else releases → true. */
    fun handle(sessionId: String, holder: String, action: String): Boolean {
        if (holder.isEmpty()) return false
        if (action == "acquire") return acquire(sessionId, holder)
        release(sessionId, holder)
        return true
    }

    companion object { const val LEASE_MS = 20_000L }
}

File: desktop/src/main/menu-answer-lock.ts
// menu-answer-lock.ts — one device at a time may answer a session's Claude Code
// menu by verified navigation (state/ink-menu-driver.ts).
//
// WHY on the HOST (review F4, 2026-09-24): an unnumbered startup dialog is
// answered with several separate keystrokes (arrows, then Enter). If the
// desktop window and a phone both clicked within ~100 ms, the desktop's arrows
// and the phone's Enter could combine into an answer neither person chose —
// e.g. trusting a folder. A per-device "sending" flag cannot see the other
// device; this lock lives in the main process, which every device's keystrokes
// pass through (desktop IPC and the remote WebSocket host share this one
// instance). The second device is refused and its card says "Another device
// is answering this".
//
// A lease, not a plain flag: a device that dies mid-answer must not lock the
// session forever. The driver's own worst case (every arrow waiting its full
// react time, then the leave wait) stays well under LEASE_MS.

export const MENU_ANSWER_LEASE_MS = 20_000;

export class MenuAnswerLock {
  private held = new Map<string, { holder: string; until: number }>();

  constructor(private now: () => number = Date.now) {}

  /** True if `holder` may answer now (the lock was free, expired, or already theirs). */
  acquire(sessionId: string, holder: string): boolean {
    const cur = this.held.get(sessionId);
    if (cur && cur.holder !== holder && cur.until > this.now()) return false;
    this.held.set(sessionId, { holder, until: this.now() + MENU_ANSWER_LEASE_MS });
    return true;
  }

  /** Free the lock — only its holder can. */
  release(sessionId: string, holder: string): void {
    if (this.held.get(sessionId)?.holder === holder) this.held.delete(sessionId);
  }

  /** The one IPC/WebSocket entry point: 'acquire' → granted?, 'release' → true. */
  handle(sessionId: string, holder: string, action: string): boolean {
    if (typeof sessionId !== 'string' || typeof holder !== 'string' || !holder) return false;
    if (action === 'acquire') return this.acquire(sessionId, holder);
    this.release(sessionId, holder);
    return true;
  }
}

/** The host's single instance — shared by ipc-handlers.ts and remote-server.ts. */
export const menuAnswerLock = new MenuAnswerLock();
